File: app/channels/media_stream_channel.rb
class MediaStreamChannel < ApplicationCable::Channel
  def subscribed
    # Stream for a specific session/call ID
    stream_from "media_stream_#{params[:call_id]}"
  end

  def receive(data)
    # Receive binary base64 audio chunks from clients
    payload_chunk = data["audio_chunk"]

    # Save to DB / track request
    db_request = DatabaseRequest.create!(
      requestable_type: "VoipChunk",
      payload: { call_id: params[:call_id], chunk: payload_chunk }.to_json,
      occurred_at: Time.current
    )

    # Re-broadcast chunk to recipient in real-time
    ActionCable.server.broadcast("media_stream_#{params[:call_id]}", {
      sender_id: current_user&.id,
      audio_chunk: payload_chunk
    })
  end
end

File: app/javascript/channels/media_stream_channel.js
// app/javascript/channels/media_stream_channel.js
import consumer from "./consumer"

const callId = "call-12345";

const mediaStreamChannel = consumer.subscriptions.create(
  { channel: "MediaStreamChannel", call_id: callId },
  {
    received(data) {
      // Play incoming audio chunk on the receiver's end
      playAudioChunk(data.audio_chunk);
    },

    sendAudioChunk(base64Chunk) {
      this.perform("receive", { audio_chunk: base64Chunk });
    }
  }
);